library/widget/basic.lean

```diff
@@ -0,0 +1,214 @@
+This process is called __rendering__.
+
+The client then displays this UI and the user may interact with it.
+If the user clicks something, the client will send a message to the widget indicating that an action `a` was performed.
+`update p s a` will then be called on the relevant subcomponent
+resulting in a pair `⟨s,b⟩ : State × option Action`.
+
+The component's state is replaced with `s`.
+If `b` is `some x`,  the parent component's `update` will also be called with `x` and so on up the component tree.
+
+Finally, the entire component is re-rendered to produce a new piece of html to send to the client for display.
+On this rerendering, the new html and the old html are compared through a process called __reconciliation__.
+Reconciliation will make sure that the states are carried over correctly and will also not rerender subcomponents if they haven't changed their props or state.
+The reconciliation engine uses the `props_eq` predicate passed to the component constructor to determine whether the props have changed and hence whether the component should be re-rendered.
+
+## Keys
+
+If you have some list of components and the list changes according to some state, it is important to add keys to the components so
+that if two components change order in the list their states are preserved.
+If you don't provide keys or there are duplicate keys then you may get some strange behaviour in both the Lean widget engine and react.
+
+## Further work:
+
+Some things that would be interesting to implement and which may improve performance.
+
+- instead of sending a whole html-tree each time, generate a patch that react can use to update
+- how should style sheets work? right now you need to just use the styles provided by the client or use inline styles.
+- timers, right now the round trip from client to server is too
+- a top level action that allows the user to
+- if `decidable_eq Props`, then reconciliation can compare the old and new props and check whether the given subcomponent needs to be re-rendered, possibly increasing efficiency.
+- a parser from jsx-like expressions?!
+- currently component and html both have to be implemented natively for universe issues, it would be really cool if they could be non-meta and expressed as lean inductive datatypes.
+
+-/
+
+namespace widget
+
+inductive mouse_event_kind
+| on_click
+| on_mouse_enter
+| on_mouse_leave
+
+meta mutual inductive component, html, attr
+
+with component : Type → Type → Type
+|mk {Props : Type}
+    {Action : Type}
+    (InnerAction : Type)
+    (State : Type)
+    (init : Props → option State → State)
+    (update : Props → State → InnerAction → State × option Action)
+    (view : Props → State → list (html InnerAction))
+    /- If this returns true, then the component will not call 'view' again. -/
+    (props_eq : Props → Props → bool)
```

Comment:
Is there a reason not to make this `decidable_eq Props`?

Reply:
Yes, sometimes you just want it to return false all the time if you want the component to always update on a props change.

Reply:
There are some variants of make component which use decidable_eq instead

Reply:
Ok, seems reasonable.